client/core/core.go

```diff
@@ -10002,3 +10002,65 @@ func (c *Core) saveDisabledRateSources() {
 		c.log.Errorf("Unable to save disabled fiat rate source to database: %v", err)
 	}
 }
+
+func (c *Core) shieldedWallet(assetID uint32) (asset.ShieldedWallet, error) {
+	w, found := c.wallet(assetID)
+	if !found {
+		return nil, fmt.Errorf("no %s wallet", unbip(assetID))
+	}
+	sw, is := w.Wallet.(asset.ShieldedWallet)
+	if !is {
+		return nil, fmt.Errorf("%s wallet is not a shielded wallet", unbip(assetID))
+	}
+	return sw, nil
+}
+
+// ShieldedStatus is the shielded balance and last address associated with the
+// shielded wallet.
+func (c *Core) ShieldedStatus(assetID uint32) (*asset.ShieldedStatus, error) {
+	sw, err := c.shieldedWallet(assetID)
+	if err != nil {
+		return nil, err
+	}
+	return sw.ShieldedStatus()
+}
+
+// NewShieldedAddress creates a new shielded address. Visit the link below for
+// additional notes on shielded address reuse.
+// https://electriccoin.co/blog/shielded-address-contexts/
+func (c *Core) NewShieldedAddress(assetID uint32) (string, error) {
+	sw, err := c.shieldedWallet(assetID)
+	if err != nil {
+		return "", err
+	}
+	return sw.NewShieldedAddress()
+}
+
+// ShieldFunds moves funds from the transparent account to the shielded account.
+func (c *Core) ShieldFunds(assetID uint32, amt uint64) (dex.Bytes, error) {
+	sw, err := c.shieldedWallet(assetID)
```

Comment:
These functions should take a password.

Reply:
Why?  webserver is authenticated

Reply:
True, I just thought because `Send` also requires a password. `Send` uses it to unlock the wallet, I guess that could be done over here as well. Also, the RPCServer doesn't check the password anywhere.

Reply:
It probably makes sense for `SendShielded`, since that can send to any address. We require a password for `(*Core).Send`.

Reply:
The rpc server is authenticated like the webserver.  But unlocking the wallet is a good reason

Reply:
> The rpc server is authenticated like the webserver. But unlocking the wallet is a good reason

Actually, why is unlocking that wallet a good reason?  There are methods to unlock the wallet

Reply:
Zcash wallet isn't an `Authenticator` anyway, but wallets that are authenticated required unlocking for sends. But our main goal is to add an authentication step even if the user has a cached password. That's why we're not using `getCachedPasswordUsingRequest` in `apiSend`.

Reply:
Alright, makes sense.

But I've been a little confused about the Core API and why certain things require a password.  Take bitcoind for example.  You do `walletpassphrase` and then you can do `sendtoaddress`/`signrawtransaction`/etc. without entering the wallet password again.  The node's RPC server's are user/password protected.
Similarly, there is authentication on dexc's web and RPC servers.  Although in the case of webserver the endpoints requiring auth really pertain to successfully having logged in once and stored an auth cookie, while with the RPC server it just means you passed the correct RPC credentials.

Anyway, the reason why it confuses me a bit is because `Core` methods are a Go API, not an interface to the user (there's always another layer between the user and Core where permissioning requests happens).  It's not a big deal to me, I'm just not clear about it.

Put another way, it seems to me that a bunch of the `Core` methods that require a password could really just check internally an `isLoggedIn` flag.  Combined with the interface's auth (pass+cookie for webserver, or user/pass for rpcserver), that seems to have the right effect.

Reply:
We typically require a password when we need a `Crypter`, and we don't ever store a `Crypter` in memory, nor do we want to. I think what you're getting at is that for many methods, the `Crypter` is only needed to unlock wallets, so why not just have the caller unlock wallets before using those methods. That's fair. I think the reason is more because our UX goals are bleeding over into our back end design. We want the user to be able to perform all actions with a single form. If they want to trade on a locked wallet, we just add an app password field to the trade form, for instance, rather than showing them two different dialogs. Perhaps we should leave that UX intact, but have the front end send the `/openwallet` request separately. That's reasonable, I think. I don't know if it's worth an API-breaking refactor at this point, and the "password any time encryption might be needed" approach doesn't bother me so much.

Reply:
> the `Crypter` is only needed to unlock wallets, so why not just have the caller unlock wallets before using those methods. That's fair

Exactly.  And if it does not require an unlock, then we probably didn't need the password on the core method I think.  In general, Send can require an unlock, even if ZEC doesn't, so I get it.

>  I think the reason is more because our UX goals are bleeding over into our back end design

Yes, but not just backend.  It's transcending backend *layers* webserver -> Core.

> I don't know if it's worth an API-breaking refactor at this point, and the "password any time encryption might be needed" approach doesn't bother me so much.

No not really worth it.  I'm just trying to get to the heart of our policy regarding *Core*'s password requirements when there's no obvious need for a crypter.  Sometimes it seems like we're just double authorizing at two different backend layers.

Since some assets can require unlock to send, I think you've made a good case to keep it as it is, and have this shielding require a pass too, but there are other was to approach this as you've layed out.